### PR TITLE
flake-check: use stable kubevirtci provider, set 3 nodes if required

### PR DESCRIPTION
**Pull request by dhiller**
**What this PR does / why we need it**:

This PR fixes three things:

# Use a stable kubevirtci provider

We do not want to use the latest provider if it is not stable yet, since there might be regressions that are not yet fixed in an rc.

Thus we use the latest stable provider instead.

# Set 3 nodes if required

Some tests need two nodes with cpu manager installed, since the first node doesn't have it installed, they then fail.

Set nodes to 3 in that case.

# increase default node memory

The [default test script increases the node memory](https://github.com/kubevirt/kubevirt/blob/main/automation/test.sh#L109).

Since [a test was failing for this reason](https://testgrid.k8s.io/kubevirt-presubmits#pull-kubevirt-check-tests-for-flakes) (thanks @jean-edouard) we increase the node memory to the same amount.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
/cc @jean-edouard @brianmcarey 

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: automation/repeated_test.sh [hunk 1]
```diff
@@ -123,7 +129,21 @@ if (( $# > 0 )); then
         shift
     done
 else
-    mapfile -t TEST_LANES < <( find cluster-up/cluster -name 'k8s-[0-9].[0-9][0-9]' -print | sort -rV | grep -oE 'k8s.*' | head -1 )
+    # We only want to use stable providers for flake testing, thus we fetch the k8s version file from kubevirtci.
+    # we stop at the first provider that is stable (aka doesn't have an rc or beta or alpha version)
+    for k8s_provider in $(find cluster-up/cluster -name 'k8s-[0-9].[0-9][0-9]' -print | sort -rV | grep -oE 'k8s.*'); do
```
Comment (jean-edouard) [inline automation/repeated_test.sh:134]:
```suggestion
    for k8s_provider in $(cd cluster-up/cluster && ls -rd k8s-[0-9]\.[0-9][0-9]); do
```

File: automation/repeated_test.sh [hunk 2]
```diff
@@ -123,7 +129,21 @@ if (( $# > 0 )); then
...
+        k8s_provider_version=$(curl --fail "https://raw.githubusercontent.com/kubevirt/kubevirtci/${kubevirtci_git_hash}/cluster-provision/k8s/${k8s_provider#"k8s-"}/version")
+        if [[ ! "${k8s_provider_version}" =~ -(rc|alpha|beta) ]]; then
+            TEST_LANES=("${k8s_provider}")
+            break
+        else
+            echo "Skipping ${k8s_provider_version}"
+        fi
```
Comment (jean-edouard) [inline automation/repeated_test.sh:141]: nit: removing the `!` in the condition and swapping the then and the else would make this more readable IMO

File: automation/repeated_test.sh [hunk 3]
```diff
@@ -123,7 +129,21 @@ if (( $# > 0 )); then
...
     done
 else
-    mapfile -t TEST_LANES < <( find cluster-up/cluster -name 'k8s-[0-9].[0-9][0-9]' -print | sort -rV | grep -oE 'k8s.*' | head -1 )
+    # We only want to use stable providers for flake testing, thus we fetch the k8s version file from kubevirtci.
+    # we stop at the first provider that is stable (aka doesn't have an rc or beta or alpha version)
+    for k8s_provider in $(find cluster-up/cluster -name 'k8s-[0-9].[0-9][0-9]' -print | sort -rV | grep -oE 'k8s.*'); do
+        k8s_provider_version=$(curl --fail "https://raw.githubusercontent.com/kubevirt/kubevirtci/${kubevirtci_git_hash}/cluster-provision/k8s/${k8s_provider#"k8s-"}/version")
+        if [[ ! "${k8s_provider_version}" =~ -(rc|alpha|beta) ]]; then
+            TEST_LANES=("${k8s_provider}")
+            break
```
Comment (jean-edouard) [inline automation/repeated_test.sh:138]: So we're now testing only 1 test lane? If so we should probably use a simple string variable, ditch the S at the end and remove the for loop on it.

File: automation/repeated_test.sh [hunk 4]
```diff
@@ -158,17 +178,31 @@ if should_skip_test_run_due_to_too_many_tests "${NEW_TESTS}"; then
     exit 0
 fi
 
+# for migration tests we need three nodes, thus we need to check whether migration tests are going to be run
+KUBEVIRT_NUM_NODES=2
+if [[ "${NEW_TESTS}" =~ migration ]]; then
+    KUBEVIRT_NUM_NODES=3
```
Comment (jean-edouard) [inline automation/repeated_test.sh:184]: We only need 3 nodes for the tests that include a call to `SkipTestIfNotEnoughNodesWithCPUManager`, not all the migration tests!

File: automation/repeated_test.sh [hunk 5]
```diff
@@ -158,17 +178,31 @@ if should_skip_test_run_due_to_too_many_tests "${NEW_TESTS}"; then
...
+else
+    if grep -q 'SkipTestIfNotEnoughNodesWithCPUManager' $(echo ${NEW_TESTS} | tr '|' ' '); then
```
Comment (jean-edouard) [inline automation/repeated_test.sh:186]: Does `$NEW_TESTS` include the entire code of each test? That skip is a function that's called by a test, it's not part of the test entry line like a decorator would.
I'd suggest creating a new decorator to be able to easily and reliably identify the (currently 3) tests that need 3 nodes.

File: automation/repeated_test.sh (hunk 1 above)
Comment (dhiller) [reply automation/repeated_test.sh:134]: TIL: `ls -rd` sorts the directories in reverse order like `sort -rV`.

Thanks!

File: automation/repeated_test.sh (hunk 5 above)
Comment (dhiller) [reply automation/repeated_test.sh:186]: No, `NEW_TESTS` only contains the file names of the touched test files.

File: automation/repeated_test.sh (hunk 1 above)
Comment (jean-edouard) [reply automation/repeated_test.sh]: With the caveat that `ls` does an alphabetical sort, not a version sort. So if/when we get to k8s version `1.100` or `2.10`, they'll be considered older than resp. `1.99` and `2.9`.
If you want to be that future-proof, you might want to still do `ls -d | sort -rV`.

**Comment by dhiller** [comment]
@jean-edouard thanks for the review, I've addressed the comments. PTAL.

**Review by jean-edouard — COMMENTED** [review]
/approve

**Review by brianmcarey — COMMENTED** [review]
/lgtm

**Pull request merged**